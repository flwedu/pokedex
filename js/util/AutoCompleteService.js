class AutoCompleteService {
  constructor(dataPath) {
    this._arrayDeDados = [];
    this.onload(dataPath);
  }

  onload(dataPath) {
    fetch(dataPath)
      .then((res) => {
        return res.ok ? res.json() : Promise.reject(res.status);
      })
      .then((res) => this.extrairPropertyParaArray(res))
      .catch(console.error);
  }

  extrairPropertyParaArray(objetoBruto) {
    // Transferindo os valores do objeto para o array
    // Dos valores já são retirados os nomes e convertidos para lowercase
    this._arrayDeDados = Object.values(objetoBruto).map((element) =>
      element.name.toLowerCase()
    );
  }

  listarPorAproximacao(name) {
    // Retornando os 5 primeiros elementos do array de dados que satisfaçãm a query
    return this._arrayDeDados.filter((item) => item.includes(name)).slice(0, 5);
  }
}
