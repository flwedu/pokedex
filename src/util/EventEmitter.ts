export default class EventEmitter {
  private events;

  constructor() {
    this.events = new Map<string, Function[]>();
  }

  on(topic: string, callback: Function) {
    const oldEvents = this.events.get(topic);
    if (this.events.has(topic)) {
      this.events.set(topic, [...oldEvents, callback]);
    }
    return this.events.set(topic, [callback]);
  }

  emit(topic: string, data: any) {
    const listeners = this.events.get(topic);
    if (Array.isArray(listeners) && listeners.length) {
      listeners.forEach((event) => event(data));
    }
  }
}
