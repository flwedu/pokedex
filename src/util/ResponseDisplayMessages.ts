import { IPokemon } from "../model/Pokemon";
import formatText from "./text";

const errorImageSrc = "../../assets/error.svg";

const errorMessages = new Map();
errorMessages.set("default", "Error");
errorMessages.set("404", "No Pokémon found");
errorMessages.set("500", "Server Error");

/**
 * Function that returns a literal string with pokemon data to be used in the element inner text
 * @param {IPokemon} pokemon
 * @returns {string}
 */
export function pokemonData(pokemon: IPokemon): string {
  const parentElement = document.createElement("div");
  const h2 = document.createElement("h2");
  const img = document.createElement("img");
  const p = document.createElement("p");
  const pw = document.createElement("p");

  h2.textContent = `Nº:${pokemon.id} / ${formatText(pokemon.name)}`;
  img.src = pokemon.sprites.front_default;
  p.textContent = `Type: ${pokemon.types.map(type => type.type.name).join(" & ")}`;
  pw.textContent = `Weight: ${pokemon.weight} (in hectograms)`;

  parentElement.append(h2, img, p, pw);
  return parentElement.innerHTML;
}

/**
 * Function that returns a literal string with pokemon stats data to be used in the element inner text
 * @param {IPokemon} pokemon
 * @returns {string}
 */
export function pokemonStats(pokemon: IPokemon): string {
  const parentElement = document.createElement("div");
  const p = document.createElement("p");
  const ul = document.createElement("ul");

  p.textContent = "Initial Stats of Pokémon";

  ul.innerHTML = `
  ${pokemon.stats.map(stat => `
  <li>${stat.stat.name}: ${stat.base_stat}<li>
  `).join("")}
  `;

  parentElement.append(p, ul);
  return parentElement.innerHTML;
}

export function pokemonAbilities(pokemon: IPokemon) {

}

/**
 *
 * @param {Response} response
 * @returns {string}
 */
export function errorMessageWithResponseCode(response: Response): string {
  const parentElement = document.createElement("div");
  const h2 = document.createElement("h2");
  const p = document.createElement("p");
  const img = document.createElement("img");

  h2.textContent = `Error ${response.status}`;
  p.textContent = `${errorMessages.get(response.status.toString()) || errorMessages.get("default")}`;
  img.src = errorImageSrc;

  parentElement.append(h2, p, img);

  return parentElement.innerHTML;
}
