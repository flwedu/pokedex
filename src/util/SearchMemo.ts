import { IPokemon } from "src/model/Pokemon";

export class SearchMemo {
  private readonly memo: Map<string, IPokemon>;

  constructor() {
    this.memo = new Map();
  }

  public add(query: string, pokemon: IPokemon) {
    this.memo.set(query, pokemon);
  }

  public get(query: string): IPokemon | undefined {
    return this.memo.get(query);
  }

  public has(query: string): boolean {
    return this.memo.has(query);
  }

  public clear() {
    this.memo.clear();
  }
}
