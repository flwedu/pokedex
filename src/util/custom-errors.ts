class NotFoundError extends Error {
	constructor(message: string = "Pok√©mon not found") {
		super(message);
		this.name = "NotFoundError";
	}
}

class BadRequestError extends Error {
	constructor(message: string = "Bad request") {
		super(message);
		this.name = "BadRequestError";
	}
}

class InternalServerError extends Error {
	constructor(message: string = "Internal server error") {
		super(message);
		this.name = "InternalServerError";
	}
}

const statusErrors: Record<number, Error> = {
	400: new BadRequestError(),
	404: new NotFoundError(),
	500: new InternalServerError(),
};

export { NotFoundError, BadRequestError, InternalServerError, statusErrors };
