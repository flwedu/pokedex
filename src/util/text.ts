export function formatFirstLetterToUppercase(texto: string) {
	const [firstLetter, ...rest] = texto;
	return rest.reduce(
		(prev, next) => prev + next.toLowerCase(),
		firstLetter.toUpperCase()
	);
}

export function removeInvalidCharacters(text: string) {
	const number = Number.parseInt(text);
	if (!Number.isNaN(number)) return number;

	const symbolRegex = /[^a-zA-Z]/g;
	return text.toLowerCase().replace(symbolRegex, "");
}
