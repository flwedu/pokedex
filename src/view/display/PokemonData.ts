import { IPokemon } from "../../model/Pokemon";
import formatText from "../../util/text";

export function pokemonData(pokemon: IPokemon): string {
    const parentElement = document.createElement("div");
    const h2 = document.createElement("h2");
    const img = document.createElement("img");
    const p = document.createElement("p");
    const pw = document.createElement("p");

    h2.textContent = `NÂº:${pokemon.id} / ${formatText(pokemon.name)}`;
    img.src = pokemon.sprites.front_default;
    p.textContent = `Type: ${pokemon.types.map(type => type.type.name).join(" & ")}`;
    pw.textContent = `Weight: ${pokemon.weight} (in hectograms)`;

    parentElement.append(h2, img, p, pw);
    return parentElement.innerHTML;
}