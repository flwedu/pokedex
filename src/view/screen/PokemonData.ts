import IScreen from ".";
import {IPokemon} from "../../model/Pokemon";
import {formatFirstLetterToUppercase} from "../../util/text";

export class PokemonData implements IScreen {
  getHtml(pokemon: IPokemon): string {
	  return `
  <div class="container">
  <h2>NÂº:${pokemon.id} / ${formatFirstLetterToUppercase(pokemon.name)}</h2>
  <img src="${pokemon.sprites.front_default}" alt="${pokemon.name} picture">
  <p>Type: ${pokemon.types.map((type) => type.type.name).join(" & ")}</p>
  <p>Weight: ${pokemon.weight} (in hectograms)</p>
  </div>
  `;
  }
}
