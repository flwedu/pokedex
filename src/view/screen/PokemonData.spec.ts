import { ditto } from "../../util/mock_data";
import { pokemonData } from "./PokemonData";

describe("Pokemon data screen", () => {
  it("should render the pokemon data", () => {
    const html = pokemonData(ditto);

    expect(html).toMatchSnapshot();
  });
});
