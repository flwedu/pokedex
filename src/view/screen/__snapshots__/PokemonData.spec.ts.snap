// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon data screen should render the pokemon data 1`] = `
"
  <div>
  <h2>NÂº:132 / Ditto</h2>
  <img src=\\"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/132.png\\" alt=\\"ditto picture\\">
  <p>Type: normal</p>
  <p>Weight: 40 (in hectograms)</p>
  </div>
  "
`;
